src/solver/batchnorm/forward_inference_ck.cpp

```diff
@@ -126,52 +126,60 @@ template <typename XDataType,
-        ck::Tuple<XDataType, MeanVarDataType, MeanVarDataType, ScaleDataType, BiasDataType>,
-        ck::Tuple<YDataType>,
-        Normalize,
-        Rank>;
-    const auto bn_fwd_ptrs = ck::tensor_operation::device::instance::DeviceOperationInstanceFactory<
-        DeviceOp>::GetInstances();
+    ConvSolution result;
+    result.invoker_factory = [args         = CKArgsBNormFwd{bn_problem},
+                              kernel_index = CheckCKApplicability<XDataType,
+                                                                  YDataType,
+                                                                  AccDataType,
+                                                                  ScaleDataType,
+                                                                  BiasDataType,
+                                                                  MeanVarDataType>(bn_problem)](
+                                 const std::vector<Kernel>& /*kernels*/) mutable {
+        return [=, args = std::move(args)](const Handle& handle,
+                                           const AnyInvokeParams& primitive_parameters) {
+            using DeviceOp = ck::tensor_operation::device::DeviceElementwise<
+                ck::Tuple<XDataType, MeanVarDataType, MeanVarDataType, ScaleDataType, BiasDataType>,
+                ck::Tuple<YDataType>,
+                Normalize,
+                Rank>;
+            const auto bn_fwd_ptrs =
+                ck::tensor_operation::device::instance::DeviceOperationInstanceFactory<
+                    DeviceOp>::GetInstances();
 
-    int kernel_index = CheckCKApplicability<XDataType,
-                                            YDataType,
-                                            AccDataType,
-                                            ScaleDataType,
-                                            BiasDataType,
-                                            MeanVarDataType>(problem);
-    assert(kernel_index >= 0 && kernel_index < bn_fwd_ptrs.size());
-    auto& bn_ptr       = bn_fwd_ptrs.at(kernel_index);
-    const auto& params = primitive_parameters.CastTo<miopen::batchnorm::InfInvokeParams>();
+            assert(kernel_index >= 0 && kernel_index < bn_fwd_ptrs.size());
+            auto& bn_ptr       = bn_fwd_ptrs.at(kernel_index);
+            const auto& params = primitive_parameters.CastTo<miopen::batchnorm::InfInvokeParams>();
 
-    auto argument_ptr = bn_ptr->MakeArgumentPointer(
-        args.xyLengths,
-        {args.xyStrides,
-         args.aligned_scaleBiasMeanVarStrides,
-         args.aligned_scaleBiasMeanVarStrides,
-         args.aligned_scaleBiasMeanVarStrides,
-         args.aligned_scaleBiasMeanVarStrides},
-        {args.xyStrides},
-        {params.x, params.estimatedMean, params.estimatedVariance, params.bnScale, params.bnBias},
-        {params.y},
-        Normalize{params.epsilon});
+            auto argument_ptr = bn_ptr->MakeArgumentPointer(args.xyLengths,
```

Comment:
add a method to `BnArgs` called `MakeArgPtr` and you can then pick the fields from inside that class. Better design IMO for composability and hiding details.

Reply:
@amberhassaan done